feat: add plugin schema (#37)
coreruleset/plugin-registry: docs/plugin-descriptor-schema.md

| `status`           | yes      | Maturity level: `tested`, `being-tested`, `untested`, or `draft`. |
| `license`          | yes      | SPDX license identifier (e.g., `Apache-2.0`, `GPL-2.0-only`). |
| `authors`          | yes      | List of author objects with `name` (required), `email` and `url` (optional). |
| `repository`       | yes      | URL of the plugin source repository. |
| `homepage`         | no       | URL to documentation or project site. |
| `keywords`         | no       | Tags for discovery and categorization. |

### `rule_id_range`

**Required.** The registered rule ID block from the plugin registry.

| Field   | Description |
|---------|-------------|
| `start` | First rule ID in the allocated range (integer, 9500000-9999999). |
| `end`   | Last rule ID in the allocated range (integer, 9500000-9999999). |

Plugins typically receive a 1,000-ID range. The schema validates that values fall within the CRS plugin namespace (9,500,000 - 9,999,999).

### `compatibility`

**Optional.** WAF engine and CRS version requirements.

| Field         | Description |
|---------------|-------------|
| `crs_version` | Version constraint string (e.g., `>=4.0.0`). |

> | `engines`     | List of compatible WAF engines, each one of: `modsecurity2`, `modsecurity3`, or `coraza`. |

Reviewer: ```suggestion
| `engines`     | List of compatible WAF engines, one of: `modsecurity2`, `modsecurity3`, `coraza`, `all`. |
```

> When omitted, no compatibility constraints are assumed. Tooling should treat missing engines as "compatible with all".

Reviewer: ```suggestion
When omitted, no compatibility constraints are assumed. Tooling should treat missing `engines` field as if `engines` were set tp `all`.
```

| Field       | Description |
|-------------|-------------|
| `file`      | Path to the `-config.conf` file relative to the plugin root. |
| `variables` | List of transaction variable definitions (see below). |

#### Variable Definition

Each entry in `variables` describes a single `tx.*` variable from the config file:

| Field            | Required | Description |
|------------------|----------|-------------|
| `name`           | yes      | Full ModSecurity variable name (e.g., `tx.myplugin_enabled`). |
| `type`           | yes      | Data type: `boolean`, `integer`, `string`, `list`, or `enum`. |
| `default`        | no       | Default value if not set by the user. |
| `description`    | yes      | Human-readable explanation of the variable. |
| `required`       | no       | Whether the user must explicitly set this variable (default: `false`). |
| `allowed_values` | no       | List of valid values when type is `enum`. |
| `separator`           | no       | String used to separate multiple entries when type is `list`. |
| `prefix`                 | no       | String marking the beginning of a list entry when type is `list`. |

> | `suffix`                 | no       | String marking the end of a list entry when type is `list`. |

Reviewer: Fix table formatting.

1. Render appropriate input widgets (toggle for boolean, slider for bounded integer, dropdown for enum).
2. Validate user input before generating configuration.
3. Generate documentation automatically.

### Why is `configuration` required?

Every CRS plugin has at least one configuration variable (`tx.<name>_enabled`). Making this section required ensures consistency and guarantees that tooling always has something to work with, even for minimal rule-exclusion plugins.

### Why no `version` field?

The plugin version is derived from GitHub release tags at query time. Embedding a version in `plugin.yaml` would inevitably drift because developers forget to bump it. Tooling should fetch the latest release tag from the repository URL instead. This keeps `plugin.yaml` as a static descriptor that rarely needs updating.

### Why separate `rule_id_range` from `plugin`?

The rule ID range is a registry-level concern (namespace coordination), not a plugin identity attribute. Keeping it separate makes it clear that this range is allocated by the registry and should not be changed unilaterally.

> ### Category taxonomy

Reviewer: Add a similar section for engines.

# CRS Plugin Descriptor Schema

## Overview

The plugin descriptor schema defines a `plugin.yaml` file that lives in the root of each CRS plugin repository. It provides machine-readable metadata about the plugin, its configuration variables, and compatibility requirements.

The plugin registry aggregates these descriptors to generate the registry table, and downstream tooling (such as a CRS configurator) can parse them to build preconfigured CRS deployments based on plugin selection.

## Goals

- Allow each plugin repository to be the single source of truth for its own metadata.
- Enable automated tooling to discover, validate, and configure plugins.
- Replace manual registry table maintenance with generated output.
- Provide enough information for a configurator to present a UI for plugin selection and variable tuning.

## Schema Structure

> The schema is defined as JSON Schema (2020-12) in [`plugin-schema.json`](../plugin-schema.json). A `plugin.yaml` file has the following top-level sections:

Reviewer: ```suggestion
The schema is defined as [JSON Schema (2020-12)](https://json-schema.org/draft/2020-12/json-schema-core) in [`plugin-schema.json`](../plugin-schema.json). A `plugin.yaml` file has the following top-level sections:
```

| `status`           | yes      | Maturity level: `tested`, `being-tested`, `untested`, or `draft`. |
| `license`          | yes      | SPDX license identifier (e.g., `Apache-2.0`, `GPL-2.0-only`). |
| `authors`          | yes      | List of author objects with `name` (required), `email` and `url` (optional). |
| `repository`       | yes      | URL of the plugin source repository. |
| `homepage`         | no       | URL to documentation or project site. |
| `keywords`         | no       | Tags for discovery and categorization. |

### `rule_id_range`

**Required.** The registered rule ID block from the plugin registry.

| Field   | Description |
|---------|-------------|
| `start` | First rule ID in the allocated range (integer, 9500000-9999999). |
| `end`   | Last rule ID in the allocated range (integer, 9500000-9999999). |

Plugins typically receive a 1,000-ID range. The schema validates that values fall within the CRS plugin namespace (9,500,000 - 9,999,999).

### `compatibility`

**Optional.** WAF engine and CRS version requirements.

| Field         | Description |
|---------------|-------------|
| `crs_version` | Version constraint string (e.g., `>=4.0.0`). |

> | `engines`     | List of compatible WAF engines: `modsecurity2`, `modsecurity3`, `coraza`. |

Reviewer: Categories are an enum, engines should probably be one as well.

When omitted, no compatibility constraints are assumed. Tooling should treat missing engines as "compatible with all".

### `dependencies`

**Optional.** List of other CRS plugins this plugin depends on.

Each entry has:
| Field     | Required | Description |
|-----------|----------|-------------|
| `name`    | yes      | Name of the required plugin. |
| `version` | no       | Version constraint for the dependency. |

Most plugins have no dependencies. This field exists for plugins that build on top of other plugins.

### `configuration`

**Required.** The key section for automated tooling. Describes the config file and all user-tunable variables.

| Field       | Description |
|-------------|-------------|
| `file`      | Path to the `-config.conf` file relative to the plugin root. |
| `variables` | List of transaction variable definitions (see below). |

#### Variable Definition

Each entry in `variables` describes a single `tx.*` variable from the config file:

| Field            | Required | Description |
|------------------|----------|-------------|
| `name`           | yes      | Full ModSecurity variable name (e.g., `tx.myplugin_enabled`). |

> | `type`           | yes      | Data type: `boolean`, `integer`, `string`, or `enum`. |

Reviewer: ```suggestion
| `type`           | yes      | Data type: `boolean`, `integer`, `string`, `list`, or `enum`. |
```

| `default`        | no       | Default value if not set by the user. |
| `description`    | yes      | Human-readable explanation of the variable. |
| `required`       | no       | Whether the user must explicitly set this variable (default: `false`). |

> | `allowed_values` | no       | List of valid values when type is `enum`. |

Reviewer: ```suggestion
| `allowed_values` | no       | List of valid values when type is `enum`. |
| `separator`           | no       | String used to separate multiple entries when type is `list`. |
| `prefix`                 | no       | String marking the beginning of a list entry when type is `list`. |
| `suffix`                 | no       | String marking the end of a list entry when type is `list`. |
```

| `example`        | no       | Example value for documentation. |
| `min`            | no       | Minimum value when type is `integer`. |
| `max`            | no       | Maximum value when type is `integer`. |

#### Variable Types

The four types cover all patterns found across existing CRS plugins:

- **`boolean`** — Enable/disable flags. Every plugin has at least `tx.<name>_enabled`. Values are integers: `0` (disabled) or `1` (enabled), following the ModSecurity convention for boolean flags.
- **`integer`** — Numeric thresholds and limits (e.g., `tx.body-decompress-plugin_max_data_size_bytes`). Supports `min`/`max` constraints.

> - **`string`** — Freeform text values (e.g., `tx.google-oauth2-plugin_whitelisted_parameters`). The `example` field helps users understand the expected format.

Reviewer: ```suggestion
- **`string`** — Freeform text values. The `example` field helps users understand the expected format.
- **`list`** — Structured list of strings (e.g., `tx.google-oauth2-plugin_whitelisted_parameters`). The fields `separator`, `prefix`, `suffix` define the format. For example, the list `|a/ |b/ |c/` would be defined as follows:
  - `separator`: `" "` (space character)
  - `prefix`: `"|"`
  - `suffix`: `"/"`
```

- **`enum`** — Constrained choices (e.g., `tx.phpmyadmin-rule-exclusions-plugin_url_format` with values `v51`, `v52`). Must include `allowed_values`.

### `files`

**Optional.** Maps the standard plugin files for tooling convenience.

| Field    | Description |
|----------|-------------|

> | `config` | Path to the configuration file. |

Reviewer: Can be derived from the tooling section.

### `schema_version`

**Required.** Integer, currently fixed at `1`.

Allows future evolution of the schema without breaking existing parsers. Tooling should check this field and handle unknown versions gracefully.

### `plugin`

**Required.** Plugin identity and metadata.

| Field              | Required | Description |
|--------------------|----------|-------------|
| `name`             | yes      | Plugin name following the CRS convention: `<name>-plugin`. Validated by regex. |
| `description`      | yes      | One-line summary of the plugin's purpose. |
| `long_description` | no       | Multi-line extended description for documentation and UI display. |
| `type`             | yes      | `official` (coreruleset-maintained) or `3rd-party`. |
| `category`         | no       | Functional category: `rule-exclusion`, `detection`, `protection`, `utility`, `logging`, or `performance`. |
| `status`           | yes      | Maturity level: `tested`, `being-tested`, `untested`, or `draft`. |
| `license`          | yes      | SPDX license identifier (e.g., `Apache-2.0`, `GPL-2.0-only`). |

> | `authors`          | yes      | List of author objects with `name` (required), `email` and `url` (optional). |

Reviewer 2: ```suggestion
| `authors`          | yes      | List of author objects with `name` (required), `email` (required) and `url` (optional). |
```

Author: Hmmm, this might be problematic if authors don't want to have a public way of contacting them, as spam might get into their inboxes, and/or other problems. Do we want to `require` it?

Reviewer 2: In that case add `(optional)`. I was pointing to a different type of issue - `email` is missing information if it's required of not.

| `repository`       | yes      | URL of the plugin GitHub repository (only GitHub repositories are currently supported). |
| `homepage`         | no       | URL to documentation or project site. |

> | `keywords`         | no       | Tags for discovery and categorization. |

Reviewer 2: In what format?

Author: What do you mean? String? List?

Reviewer 2: In what format the tags should be specified? Space delimited?

Reviewer: It's a YAML list, e.g.:
```yaml
keywords:
  - some
  - keyword
 ```

### `rule_id_range`

**Required.** The registered rule ID block from the plugin registry.

| Field   | Description |
|---------|-------------|
| `start` | First rule ID in the allocated range (integer, 9500000-9999999). |
| `end`   | Last rule ID in the allocated range (integer, 9500000-9999999). |

Plugins typically receive a 1,000-ID range. The schema validates that values fall within the CRS plugin namespace (9,500,000 - 9,999,999).

### `compatibility`

**Optional.** WAF engine and CRS version requirements.

| Field         | Description |
|---------------|-------------|
| `crs_version` | Version constraint string (e.g., `>=4.0.0`). |
| `engines`     | List of compatible WAF engines, one of: `modsecurity2`, `modsecurity3`, `coraza`, `all`. |

> When omitted, no compatibility constraints are assumed. Tooling should treat missing `engines` field as if `engines` were set tp `all`.

Reviewer: ```suggestion
When omitted, no compatibility constraints are assumed. Tooling should treat missing `engines` field as if `engines` were set to `all`.
```

### `dependencies`

**Optional.** List of other CRS plugins this plugin depends on.

Each entry has:
| Field     | Required | Description |
|-----------|----------|-------------|
| `name`    | yes      | Name of the required plugin. |
| `version` | no       | Version constraint for the dependency. |

Most plugins have no dependencies. This field exists for plugins that build on top of other plugins.

### `configuration`

**Required.** The key section for automated tooling. Describes the config file and all user-tunable variables.

| Field       | Description |
|-------------|-------------|
| `file`      | Path to the `-config.conf` file relative to the plugin root. |

> | `variables` | List of transaction variable definitions (see below). |

Reviewer 2: In what format?

Reviewer: It's YAML. "List" means list, e.g.,
```yaml
variables:
 - var1
 - var2
```

#### Variable Definition

Each entry in `variables` describes a single `tx.*` variable from the config file:

| Field            | Required | Description |
|------------------|----------|-------------|
| `name`           | yes      | Full ModSecurity variable name (e.g., `tx.myplugin_enabled`). |
| `type`           | yes      | Data type: `boolean`, `integer`, `string`, `list`, or `enum`. |
| `default`        | no       | Default value if not set by the user. |
| `description`    | yes      | Human-readable explanation of the variable. |
| `required`       | no       | Whether the user must explicitly set this variable (default: `false`). |

> | `allowed_values` | no       | List of valid values when type is `enum`. |

Reviewer 2: Should be required for `enum`.

> | `separator`      | no       | String used to separate multiple entries when type is `list`. |

Reviewer 2: Should be required for `list`.

> | `prefix`         | no       | String marking the beginning of a list entry when type is `list`. |

Reviewer 2: ```suggestion
| `prefix`         | no       | String marking the beginning of a list entry when type is `list` (see below). |
```

> | `suffix`         | no       | String marking the end of a list entry when type is `list`. |

Reviewer 2: ```suggestion
| `suffix`         | no       | String marking the end of a list entry when type is `list` (see below). |
```

| `example`        | no       | Example value for documentation. |
| `min`            | no       | Minimum value when type is `integer`. |
| `max`            | no       | Maximum value when type is `integer`. |

#### Variable Types

> The four types cover all patterns found across existing CRS plugins:

Reviewer 2: I'm using another type of variables in few of 'my' plugins - it's some kind of an array, used in Lua plugins. For example:
```
tx.false-positive-report-plugin_smtp_cc_1=first@example.com
tx.false-positive-report-plugin_smtp_cc_2=second@example.com
tx.false-positive-report-plugin_smtp_cc_3=third@example.com
```

Author: Can you point me to the code to take a look?

Reviewer 2: https://github.com/coreruleset/false-positive-report-plugin/blob/b33cfa374e9563a4bc46b9747374ed79d2b0ae81/plugins/false-positive-report.lua#L191

Author: Thanks, that Lua loop is what I needed. Since each entry is its own `tx.*` variable rather than one delimited value, `list` cannot describe it, so `f654fcb` adds a sixth variable type for it:

```yaml
- name: "tx.false-positive-report-plugin_smtp_cc"
  type: "indexed"
  max_index: 5
  description: "Additional CC recipients, read as _1 through _5 until one is unset"
```

`name` holds the base name without the numeric suffix and `max_index` is the highest suffix the plugin reads, so a configurator can emit `<name>_1` … `<name>_<n>` for as many entries as the user supplies and stop at your bound of 5. The schema requires `max_index` for `indexed`, forbids it for every other type, and rejects a `name` ending in `_<digits>` (writing `..._smtp_cc_1` is the mistake I expect authors to make).

Does that match what the plugin actually does? If the bound is really "until unset, no fixed maximum" in some of your plugins, say so and I will make `max_index` optional instead of required.
